File: src/day_2_part_1.rs
use fancy_regex::Regex;
use std::fs;

pub fn main(scenario: &String, input_root: String) {
    let input_file = input_root + "/day_2_" + scenario + ".txt";
    let input = fs::read_to_string(input_file).expect("Failed to read input file");

    let mut invalid_ids: Vec<i64> = Vec::new();

    let start_stop_strs = input.split(",");
    // using a regex backreference here is really pretty slow, but it works
    let re = Regex::new(r"^(\d+)\1$").unwrap();

    for start_stop in start_stop_strs {
        let start_stop_pair: Vec<&str> = start_stop.split("-").collect();
        let start: i64 = start_stop_pair[0].parse().unwrap();
        let stop: i64 = start_stop_pair[1].parse().unwrap();
        for ind in start..=stop {
            let num_str = ind.to_string();
            if re.is_match(num_str.as_str()).expect("Regex failed") {
                invalid_ids.push(ind);
            }
        }
    }
    println!("The password is {}", invalid_ids.iter().sum::<i64>());
}

File: python/day_2_part_1.py
from core import read_file, parse_inputs


def solution(scenario: str):
    input_file = scenario + ".txt"
    input = read_file(input_file)

    # parse input string
    start_stop_strs = [x.split('-') for x in input.split(',')]
    # look for invalid ids in the range defined by each start/stop string
    invalid_ids = []
    for start_stop_str in start_stop_strs:
        start, stop = [int(n) for n in start_stop_str]
        for id in range(start, stop + 1):
            id_str = str(id)
            str_length = len(id_str)
            if str_length % 2 != 0:
                continue
            midpoint = str_length // 2
            if id_str[0:midpoint] == id_str[midpoint:]:
                invalid_ids.append(id)

    print(f"password is {sum(invalid_ids)}")


if __name__ == "__main__":
    scenario = parse_inputs("day_2_test_1")
    solution(scenario)
